File: backend/src/main/java/com/zenpay/application/service/CardService.java
package com.zenpay.application.service;

import com.zenpay.application.dto.CardLimitRequest;
import com.zenpay.application.dto.CardResponse;
import com.zenpay.domain.model.Card;
import com.zenpay.domain.model.CardStatus;
import com.zenpay.domain.repository.CardRepository;
import com.zenpay.infrastructure.exception.BusinessException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CardService {

    private final CardRepository cardRepository;

    public List<CardResponse> getCards(UUID userId) {
        return cardRepository.findByUserId(userId).stream()
                .map(this::toCardResponse)
                .collect(Collectors.toList());
    }

    public CardResponse getCardById(UUID cardId) {
        Card card = findCardById(cardId);
        return toCardResponse(card);
    }

    @Transactional
    public CardResponse blockCard(UUID cardId) {
        Card card = findCardById(cardId);
        if (card.getStatus() == CardStatus.CANCELLED) {
            throw new BusinessException("CARD_CANCELLED", "Cannot block a cancelled card");
        }
        card.setStatus(CardStatus.BLOCKED);
        card = cardRepository.save(card);
        return toCardResponse(card);
    }

    @Transactional
    public CardResponse unblockCard(UUID cardId) {
        Card card = findCardById(cardId);
        if (card.getStatus() != CardStatus.BLOCKED) {
            throw new BusinessException("CARD_NOT_BLOCKED", "Card is not blocked");
        }
        card.setStatus(CardStatus.ACTIVE);
        card = cardRepository.save(card);
        return toCardResponse(card);
    }

    @Transactional
    public CardResponse adjustLimit(UUID cardId, CardLimitRequest request) {
        Card card = findCardById(cardId);
        if (card.getCreditLimit() == null) {
            throw new BusinessException("NO_CREDIT_CARD", "This card does not have a credit limit");
        }
        card.setCreditLimit(request.creditLimit());
        card.setAvailableLimit(request.creditLimit().subtract(card.getUsedLimit()));
        card = cardRepository.save(card);
        return toCardResponse(card);
    }

    private Card findCardById(UUID cardId) {
        return cardRepository.findById(cardId)
                .orElseThrow(() -> new BusinessException("CARD_NOT_FOUND", "Card not found"));
    }

    private CardResponse toCardResponse(Card card) {
        return new CardResponse(
                card.getId(), card.getCardType(), card.getStatus(), card.getCardNumber(),
                card.getCardHolderName(), card.getExpirationDate(), card.getCreditLimit(),
                card.getUsedLimit(), card.getAvailableLimit(), card.getCurrentBalance(),
                card.getPaymentDate(), card.getCutoffDate(), card.getIsVirtual(), card.getIssuedAt());
    }
}

File: src/app/shared/services/card.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Card, CardLimitRequest } from '../models/card.model';

@Injectable({ providedIn: 'root' })
export class CardService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/cards`;

  getCards(): Observable<Card[]> {
    return this.http.get<Card[]>(this.apiUrl);
  }

  getCardById(id: string): Observable<Card> {
    return this.http.get<Card>(`${this.apiUrl}/${id}`);
  }

  blockCard(id: string): Observable<Card> {
    return this.http.patch<Card>(`${this.apiUrl}/${id}/block`, {});
  }

  unblockCard(id: string): Observable<Card> {
    return this.http.patch<Card>(`${this.apiUrl}/${id}/unblock`, {});
  }

  adjustLimit(id: string, request: CardLimitRequest): Observable<Card> {
    return this.http.patch<Card>(`${this.apiUrl}/${id}/limit`, request);
  }
}
